torch/_dynamo/optimizations/distributed.py

```diff
@@ -236,42 +239,61 @@ def forward(self, *args):
                         return x
 
                 unwrap_singleton_tuple = False
-                for sn in submod.graph.nodes:
+                for sn in input_mod.graph.nodes:
                     if sn.op == "output":
                         if not isinstance(sn.args[0], tuple):
                             unwrap_singleton_tuple = True
                             sn.args = (sn.args,)
-                submod.recompile()
 
+                input_mod.recompile()
                 wrapper = WrapperModule(
-                    self.compiler(submod, args),
+                    self.compiler(input_mod, args),
                     unwrap_singleton_tuple,
                 )
                 return wrapper
 
             def run_node(self, n: Node) -> Any:
                 with fx_traceback.append_stack_trace(n.stack_trace):
                     args, kwargs = self.fetch_args_kwargs_from_env(n)
+                    new_args = []
+                    assert fake_mode
+                    for arg in args:
+                        if isinstance(arg, torch.Tensor) and not isinstance(
+                            arg, torch._subclasses.FakeTensor
+                        ):
+                            new_args.append(fake_mode.from_tensor(arg))
```

Comment:
It would be much better if we could assume all the args are already fakeified (and we just maintain the invariant that all the intermediates are fake tensors). In particular, fakeifying a tensor can trigger the allocation of new symints, but if they're not properly associated as inputs we may not be able to instantiate those variables on subsequent runs.

tl;dr I suspect arg is always a fake tensor here

Reply:
Im not sure we can, I'm pretty sure we saw non fakes in here, but maybe that was in an intermediate state of this PR. I can check again.

Reply:
you can check this by pushing a separate PR that tightens the invariant and see if it fails or not